File: member-service/src/main/java/com/hedvig/memberservice/services/BankIdService.java
package com.hedvig.memberservice.services;

import static com.hedvig.memberservice.query.CollectType.RequestType;

import com.hedvig.external.bankID.bankidTypes.CollectResponse;
import com.hedvig.external.bankID.bankidTypes.OrderResponse;
import com.hedvig.memberservice.query.CollectRepository;
import com.hedvig.memberservice.query.CollectType;
import com.hedvig.memberservice.services.bankid.BankIdApi;
import java.io.UnsupportedEncodingException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class BankIdService {

  private final BankIdApi bankIdApi;
  private final CollectRepository collectRepository;
  private final Logger log = LoggerFactory.getLogger(BankIdService.class);

  public BankIdService(BankIdApi bankIdApi, CollectRepository collectRepository) {
    this.bankIdApi = bankIdApi;
    this.collectRepository = collectRepository;
  }

  public OrderResponse auth(Long memberId) {
    OrderResponse status = bankIdApi.auth();
    log.info(
        "Started bankId AUTH autostart:{}, reference:{}",
        status.getAutoStartToken(),
        status.getOrderRef());
    trackAuthToken(status.getOrderRef(), memberId);
    return status;
  }

  public OrderResponse sign(String ssn, String userMessage, Long memberId)
      throws UnsupportedEncodingException {
    OrderResponse status = bankIdApi.sign(ssn, userMessage);
    trackSignToken(status.getOrderRef(), memberId);
    return status;
  }

  private void trackSignToken(String referenceToken, Long memberId) {
    trackReferenceToken(referenceToken, RequestType.SIGN, memberId);
  }

  private void trackAuthToken(String referenceToken, Long memberId) {
    trackReferenceToken(referenceToken, RequestType.AUTH, memberId);
  }

  private void trackReferenceToken(String referenceToken, RequestType sign, Long memberId) {
    CollectType ct = new CollectType();
    ct.token = referenceToken;
    ct.type = sign;
    ct.memberId = memberId;
    collectRepository.save(ct);
  }

  public CollectResponse authCollect(String referenceToken) {
    return bankIdApi.authCollect(referenceToken);
  }

  public CollectResponse signCollect(String referenceToken) {
    return bankIdApi.signCollect(referenceToken);
  }
}

File: member-service/src/main/java/com/hedvig/memberservice/services/BankIdService.kt
package com.hedvig.memberservice.services

import com.hedvig.memberservice.query.CollectType.RequestType

import com.hedvig.external.bankID.bankidTypes.CollectResponse
import com.hedvig.external.bankID.bankidTypes.OrderResponse
import com.hedvig.memberservice.query.CollectRepository
import com.hedvig.memberservice.query.CollectType
import com.hedvig.memberservice.services.bankid.BankIdSOAPApi
import java.io.UnsupportedEncodingException
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class BankIdService(
    private val bankIdSOAPApi: BankIdSOAPApi,
    private val collectRepository: CollectRepository
) {
    private val log = LoggerFactory.getLogger(BankIdService::class.java)

    fun auth(memberId: Long?): OrderResponse {
        val status = bankIdSOAPApi.auth()
        log.info(
            "Started bankId AUTH autostart:{}, reference:{}",
            status.autoStartToken,
            status.orderRef)
        trackAuthToken(status.orderRef, memberId)
        return status
    }

    @Throws(UnsupportedEncodingException::class)
    fun sign(ssn: String, userMessage: String, memberId: Long?): OrderResponse {
        val status = bankIdSOAPApi.sign(ssn, userMessage)
        trackSignToken(status.orderRef, memberId)
        return status
    }

    private fun trackSignToken(referenceToken: String, memberId: Long?) {
        trackReferenceToken(referenceToken, RequestType.SIGN, memberId)
    }

    private fun trackAuthToken(referenceToken: String, memberId: Long?) {
        trackReferenceToken(referenceToken, RequestType.AUTH, memberId)
    }

    private fun trackReferenceToken(referenceToken: String, sign: RequestType, memberId: Long?) {
        val ct = CollectType()
        ct.token = referenceToken
        ct.type = sign
        ct.memberId = memberId
        collectRepository.save(ct)
    }

    fun authCollect(referenceToken: String): CollectResponse {
        return bankIdSOAPApi.authCollect(referenceToken)
    }

    fun signCollect(referenceToken: String): CollectResponse {
        return bankIdSOAPApi.signCollect(referenceToken)
    }

}